force-app/main/default/classes/EncryptionRecipes.cls

```diff
@@ -411,4 +411,25 @@ public with sharing class EncryptionRecipes {
             dataToDecryptAndCheck
         );
     }
+    /**
+     * Comparisons which involve cryptography need to be performed in constant time
+     * using specialized functions to avoid timing attack effects.
+     * https://en.wikipedia.org/wiki/Timing_attack
+     */
+    public static boolean areEqualConstantTime(Blob a, Blob b) {
+        if (a.size() != b.size()) {
```

Comment:
The `size()` comparison here defeats some of the constant time properties of the function. The early exit can be used as a way to determine the expected size of the input, e.g. assume we have

`areEqualConstantTime(<user input>, COMPARISON_STRING)`

The attack would be to start with a string of 1 character which will likely early exit. The attacker continues to append 1 character until the early exit is no longer taken. The timing chart would look like

| a.length() | time (ms) |
|--|--|
| 1 | 1 |
| 2 | 1 |
| ... | 1 |
| 10 | 100 |
| 11 | 1 |

Granted, it is expected that our inputs are the same length, so the early exit may never be taken. A potential workaround would be to do something similar to Python in the comparisons on L759 and L763. 
https://github.com/python/cpython/blob/f4c03484da59049eb62a9bf7777b963e2267d187/Modules/_operator.c#L752-L772

1. Initialize `length` to `b.length()`, assign `right` to `b`
2. If `a.length() == length`, then set `left` to `a` and continue as normal
3. Otherwise, set `left` to `b` *and* initialize `result` to `1`. This ensures that the result is always false. 

Some of the above constructs are specific to the Python implementation which uses C. These may not matter for Java and its VM. The reason that the Python implementation doesn't use `if-else` due to modern CPU architecture which attempt to predict branching behavior. This is one of the underlying issues that led to the Meltdown / Spectre which allows web content to read sensitive data outside of the browser
https://en.wikipedia.org/wiki/Meltdown_(security_vulnerability)#Background_%E2%80%93_modern_CPU_design

The Python implementation can be adopted, assuming the `length` checks are based on `blob.size()` and not `a/bChars.size()`. The original blob size should be used since it's possible for `aChars` and `bChars` to be the same size, but with different values. The iteration should still be over `a/bChars.size()`. The encoding + `getChars` should always occur regardless of the conditonal.

Suggested changes
1. Leave `bChars` uninitialized / null
2. Add the conditional checks after L429 using `b.size()`, which will then assign `bChars` and `result`

If `a.size() != b.size()`, then the loop iterates over  only `aChars` since `bChars` was set to `aChars`, but `result` will always be non-zero since it is set to `1` in the conditonal

If `a.size() == b.size()`, then the loop iterates over both `aChars` and `bChars` which are guaranteed to be the same size. There is no chance of an array out of bounds exception.

Reply:
@albarivas 

Sorry I should have mentioned. If we set a precondition of the method is that `a.size() == b.size()`, then it would be sufficient to mention that in a comment, remove the `size()` check, and say that the behavior is undefined for unequal sized input (likely an array out of bounds exception). Otherwise if we wish to support unequal sized input, then we need to perform additional changes.